source/adaptive_load/metrics_plugin_impl.cc

```diff
@@ -0,0 +1,120 @@
+#include "adaptive_load/metrics_plugin_impl.h"
+
+#include "envoy/registry/registry.h"
+
+#include "external/envoy/source/common/common/assert.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "absl/status/status.h"
+#include "absl/strings/numbers.h"
+
+namespace Nighthawk {
+
+NighthawkStatsEmulatedMetricsPlugin::NighthawkStatsEmulatedMetricsPlugin(
+    const nighthawk::client::Output& nighthawk_output) {
+  errors_ = "'global' result not found in Nighthawk output.\n";
+
+  for (const nighthawk::client::Result& result : nighthawk_output.results()) {
+    if (result.name() != "global") {
+      continue;
+    }
+    errors_ = "";
+    const int64_t actual_duration_seconds =
+        Envoy::Protobuf::util::TimeUtil::DurationToSeconds(result.execution_duration());
+    const uint32_t number_of_workers =
+        nighthawk_output.results_size() == 1 ? 1 : nighthawk_output.results_size() - 1;
+    const long total_specified = nighthawk_output.options().requests_per_second().value() *
+                                 actual_duration_seconds * number_of_workers;
+    int total_sent = -1;
+    int total_2xx = -1;
+    for (const nighthawk::client::Counter& counter : result.counters()) {
+      if (counter.name() == "benchmark.http_2xx") {
+        total_2xx = counter.value();
+      } else if (counter.name() == "upstream_rq_total") {
+        total_sent = counter.value();
+      }
+    }
+    if (total_2xx < 0) {
+      errors_ += "Counter 'benchmark.total_2xx' not found.\n";
+    }
+    if (total_sent < 0) {
+      errors_ += "Counter 'upstream_rq_total' not found.\n";
```

Comment:
I don't like having the possibility of having multiple errors stored in this one string. Could we use a string vector or some such thing, and then just combine the errors together when we need them?

Reply:
Changed to a vector.